src/ServiceClient.js

```diff
@@ -83,45 +85,70 @@ module.exports = class ServiceClient extends MicroserviceApiClient {
             const { models } = await this.connect();
+            const modelsConfig = this.config.discovery.models;
 
-            const modelSettings = this.config.discovery.models || {};
-            const exposeAllModels = Object.keys(modelSettings).length === 0;
-            // Filter definitions according to configuration.
-            // If no configuration is defined, we attach all models.
-            const modelsToDefine = models.filter((definition) => {
+            models.forEach((definition) => {
                 const { name } = definition;
-                return exposeAllModels || modelSettings[name] === true;
-            });
-
-            // define the models within the given dataSource
-            modelsToDefine.forEach((definition) => {
-                const {
-                    name,
-                    properties,
-                    relations,
-                    http,
-                    methods,
-                } = definition;
-                const model = this.dataSource.createModel(name, properties, {
-                    base: 'PersistedModel',
-                    dataSource: this.dataSource,
-                    http,
-                    methods,
-                    relations,
-                });
-                // expose the model on the app to make it shareable
-                // @todo: make this configurable
-                this.dataSource.app.model(model);
-                this._models[model.modelName] = model;
+                if (ComponentConfig.modelIsExposed(name, modelsConfig)) {
+                    const model = this._defineModel(definition, modelsConfig);
+                    this._models[name] = model;
+                }
             });
+            return this;
         } catch (error) {
             if (error instanceof ConnectionMaxDelayError) {
                 throw new DiscoveryMaxDelayError(error.message, { error });
             }
             throw error;
         }
+    }
+
+    /**
+     * Hook the model into the app.
+     *
+     * If the model is configured to be global (via isGlobal property), it is mounted into the
+     * application and avialable via app.models.ModelName (Default is true).
+     *
+     * If the model is configured to be public (via isPublic property), it is declared public
+     * and exposed on the api of the consuming service (Default is false).
+     *
+     * @param {Object} definition
+     * @param {Object} [modelsConfig] the models section of the service discovery section
+     * @return {Model} - the loopback model
+     *
+     * @private
+     */
+    _defineModel(definition, modelsConfig) {
+        const {
+            name,
+            properties,
+            relations,
+            http,
+            methods,
+        } = definition;
+
+        const isPublic = ComponentConfig.modelIsPublic(name, modelsConfig);
+
+        const model = this.dataSource.createModel(name, properties, {
+            base: 'PersistedModel',
```

Comment:
shouldn't we take the model base form the definition?

Reply:
Nope, we shouldn't! We always consume the model through a remote-datasource. Imagine if someone uses a custom base model we don't know in the consuming service.